docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+
+A single `Vector128` operation allows you to operate on: 16 (s)bytes, 8 (u)shorts, 4 (u)ints/floats, or 2 (u)longs/double(s).
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough, you should use it instead of `Vector128<T>`. To check the acceleration, use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+The size of the input also matters. It needs to be at least of the size of a single vector to be able to execute the vectorized code path. `Vector128<T>.Count` and `Vector256<T>.Count` return the number of elements of the given type T in a single vector.
+Both APIs are turned into constants by the Just-In-Time compiler (i.e. no method call is required to retrieve the information). In the case of pre-compiled code (NativeAOT), this is not true for the `IsHardwareAccelerated` property, as the required information is not available at compile time.
+
+That is why the code is very often structured like this:
+
+```cs
+void CodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (Vector256.IsHardwareAccelerated && buffer.Length >= Vector256<byte>.Count)
+    {
+        // Vector256 code path
+    }
+    else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+    {
+        // Vector128 code path
+    }
+    else
+    {
+        // non-vectorized && small inputs code path
+    }
+}
+```
+
+**Both vector types provide almost identical features**, but arm64 hardware does not support `Vector256` yet, so for the sake of simplicity we will be using `Vector128` in all examples and assuming **little endian** architecture. Which means that all examples used in this document assume that they are being executed as part of the following `if` block:
+
+```cs
+else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+{
+    // Vector128 code path
+}
+```
+
+### Testing
+
+Such a code structure requires us to **test all possible code paths**:
+
+* `Vector256` is accelerated:
+  * The input is large enough to benefit from vectorization with `Vector256`.
+  * The input is not large enough to benefit from vectorization with `Vector256`, but it can benefit from vectorization with `Vector128` (when `Vector256` is accelerated then `Vector128` and smaller vectors are also).
```

Comment:
The "when `Vector256` is accelerated then `Vector128` and smaller vectors are also" assumption is worth calling out earlier in the doc when describing IsHardwareAccelerated.